Add BufferGeometry index optimization and memory functions

gkjohnson opened the pull request:
I added `optimizeTriangleIndices` and `getMemoryUse` functions to the BufferGeometry object.

### Behavior
#### BufferGeometry.mergeVertices(tolerance = 1e-4)
Updates (or creates) the vertex attributes and index arrays with optimized, deduped versions representing the same mesh. The `tolerance` argument determines the decimal precision used when comparing and deduping vertex attributes.

#### BufferGeometryUtils.estimateMemoryUse(geometry)
Returns the memory allocated for the geometry (vertex attribute and index arrays) as bytes. Useful for seeing how much memory a model is currently taking up.

#### BufferGeometry.interleaveAttributes(attrArray)
Converts the provided normal attributes into a set of interleaved attributes that share a buffer.

#### Other
- Add `copy` and `clone` functions to InterleavedAttribute

### Uses
- Allow memory-constrained applications to release memory (on GPU and CPU) by processing geometry into a more compressed form
- Optimize geometry from loaders and file formats that don't allow triangle indices (https://github.com/mrdoob/three.js/issues/13708)
- Optimize geometry before saving it with an exporter to reduce the file format size
- Estimate how much memory a piece of geometry is using (for debugging and optimization)

### Example
https://gkjohnson.github.io/threejs-sandbox/mergeVertices/

Thanks!
Garrett

gkjohnson committed:
Add optimization and memory functions

takahirox commented:
1. Does this optimization work fine even with `.groups`?

2. Just my preference tho, `BufferGeometryUtils` would be a better place to add these methods?

gkjohnson commented:
> 1. Does this optimization work fine even with `.groups`?

It does not! Thanks for catching that. I'll rework it so each group is optimized independently. I suppose one of the complexities is that groups can have overlapping draw indices, right? If that's the case I'll probably just print a warning and not optimize.

> 2. Just my preference tho, `BufferGeometryUtils` would be a better place to add these methods?

I didn't realize that was there! I think `getMemoryUse` definitely belongs there, but the triangle optimization pairs nicely with the `toNonIndexed` function, which is in `BufferGeometry`.

Thanks @takahirox!

takahirox commented:
> I suppose one of the complexities is that groups can have overlapping draw indices, right

I think so.

> If that's the case I'll probably just print a warning and not optimize.

Perhaps it sounds reasonable. Good start.

WestLangley commented:
To summarize, you provide a method that will convert non-indexed "triangle soup" to indexed `BufferGeometry` that shares vertices when possible, reducing the memory footprint.

This is akin to `Geometry.mergeVertices()`, and requires that the user specify a `precision` value, which is used for comparing floats.

Does this PR work for interleaved BufferAttrbutes?

gkjohnson commented:
> To summarize, you provide a method that will convert non-indexed "triangle soup" to indexed BufferGeometry that shares vertices when possible, reducing the memory footprint.

That's right, but in this case it's using every attribute in the hash to make sure they're all the same before merging the vertices.

> This is akin to Geometry.mergeVertices(), and requires that the user specify a precision value, which is used for comparing floats.

I hadn't seen that function, but yeah! A precision value isn't required here -- it's optional and defaults to 3. Maybe it makes more sense for the function to be called `mergeVertices` so it's consistent with `Geometry`?

Would it make sense to expose precision as an option for `Geometry.mergeVertices`, as well?

> Does this PR work for interleaved BufferAttrbutes?

It doesn't. I hadn't encountered those before -- I can take a look at what it would take to support that. Is it okay if the process of merging vertices de-interleaves the attributes?

Thanks!

gkjohnson commented:
> No, a precision value is required. It has a default value of 3.

Ha, fair enough.

> > Is it okay if the process of merging vertices de-interleaves the attributes?
>
> Probably not.

I'll think about that a little more, then.

It also just occurred to me that a BufferAttribute can be shared across multiple pieces of geometry in the scene, so modifying them in place probably isn't okay (as I am here), is that right? It seems like the function should be creating and setting _new_ BufferAttributes, right?

WestLangley commented:
`Geometry` will be deprecated at some point -- or at least will no longer be renderable -- so converting `Geometry` to `BufferGeometry` ("triangle soup") will no longer be so much of an issue.

Where I think this PR may be useful is in merging (sharing) vertices. I do not think reducing the footprint is that big of a deal.

There will be limits to the use cases for with your approach will work. You will have to determine that yourself. :)

In the end, it remains to be seen if this PR is "useful enough".

gkjohnson committed:
Merge remote-tracking branch 'mrdoob-origin/dev' into optimize-buffer-geometry-indices

gkjohnson committed:
Rename function to 'mergeVertices', create new BufferAttribute objects, make it work with interleavedAttributes

gkjohnson committed:
Add some utility functions to BufferGeometryUtils

gkjohnson committed:
Add missing clone and copy functions to InterleavedBufferAttribute

gkjohnson committed:
Include InterleavedBufferAttribute cloning in a test

gkjohnson commented:
> `Geometry` will be deprecated at some point -- or at least will no longer be renderable -- so converting `Geometry` to `BufferGeometry` ("triangle soup") will no longer be so much of an issue.

That's good to know! I was kind of wondering if that was going to happen. Is there a roadmap or anything documenting some of the future plans for the library? I don't see anything in the wiki or docs.

> Where I think this PR may be useful is in merging (sharing) vertices. I do not think reducing the footprint is that big of a deal.
>...
>In the end, it remains to be seen if this PR is "useful enough".

I recognize that my use cases aren't necessarily common ones. I do feel that minimizing the memory footprint before exporting a model to a file is more generally useful, especially when the intent is to load that back to the web.

I'd included the function in BufferGeometry because it seemed like a natural fit along with "toNonIndexed". If the consensus is that this belongs somewhere else then I'll gladly move it! I'm just here to contribute work I'm doing otherwise to a broader community who can hopefully benefit. On that topic, is there any documentation on vision or what does or doesn't belong in the core library? Or is it more community driven?

And finally, here are some of the changes and updates I've made:

### Changes
- Merging geometry actually _does_ work with overlapping and non overlapping groups
- Function renamed to `mergeVertices`
- Function creates new `BufferAttributes` so as to not mess with other geometry that may also be using them.
- Merging vertices now works with `InterleavedBufferAttributes`, but de-interleaves them in the process. I'll need a bit more time to think about a clean way to handle that if it's critical that it not de-interleave.
- Added `interleaveAttributes` to `BufferGeometryUtils` to make an interleaved attribute from an array of attributes
- Added `getMemoryUse` to `BufferGeometryUtils`
- Added `copy`, `clone`, and modified the clone for `BufferGeometry` because it was failing to clone when interleaved buffers were used
- Updated [demo](https://gkjohnson.github.io/threejs-optimize-triangle-indices/example.html) to show the changes

Usnul commented:
I have a coupld of late API change requests.

`precision` is not an easy to use concept as it has no intuitive default. I recommend changing this to `error` or `tolerance` instead, when `error === 0` it will basically do nothing, when `error === 0.001` it will work the same way as `precision === 3` (if i understand the current logic correctly). So when i use the API i can supply `0` to have nothing done - this is a handy convention. Also, precision parameter affords more control over deduping behaviour (wider range of values, not just `0-X`) and it allows working with very large models, where one might want to set `precision` to a value greater than `1`.

I recommend changing name `getMemoryUse`. First of all, I suggest using word `compute` instead of `get` as it conveys more than retrieval. Second, I recommend `BuffersByteSize` instead of `MemoryUse`. I.e. `computeBuffersByteSize` - it's a bit less intuitive, but more explicit in terms of conveying expectations on returned value. Alternatives:  `BuffersBytesUsed` or `Attributes` instead of `Buffers`, however, I feel that `Buffers` is more clear and generic at the same time.

gkjohnson committed:
Merge pull request #2 from mrdoob/dev

Merging master

gkjohnson committed:
Merge remote-tracking branch 'origin/dev' into optimize-buffer-geometry-indices

gkjohnson committed:
Change precision to tolerance, update memory function name

gkjohnson commented:
I've update the PR, again! If this all looks good I can add documentation for the function to the BufferGeometry docs page!

### Changes

- Rename `BufferGeometryUtils.getMemoryUse` to `BufferGeometryUtils.estimateBytesUsed`
- Fix `estimateBytesUsed` function to work with interleaved buffers
- Update the `mergeVertices` function to take a `tolerance` variable rather than `precision`

gkjohnson committed:
Linting update

gkjohnson committed:
Make mergeVertices work if an index array already exists

donmccurdy commented:
This PR was helpful to me for reducing the size of files when converting an old Geometry example (`sittingBox.js`) to glTF using GLTFExporter. By default the conversion from Geometry to BufferGeometry drops indices and increases file size 5x, and the mergeVertices function here brought that back down. One suggestion, though, would be to preserve morphAttributes throughout the merge:

https://gist.github.com/donmccurdy/bf17af7acfcdd5ebc57da1200e7b1e46

gkjohnson committed:
Merge remote-tracking branch 'mrdoob-origin/dev' into optimize-buffer-geometry-indices

gkjohnson commented:
Thanks @donmccurdy! I've added your changes and am trying to test it but I'm having a hard time getting the `sittingBox.js` model to animate as a `BufferGeometry`.  The `morphAttributes` are filled out but the animation from geometry doesn't seem to be working. Is there another step to converting a morphTarget animated geometry to bufferGeometry?

Thanks!

gkjohnson commented:
This is before merging the indices. Here is an example where the Geometry is animating, but I can't get it to animate once it has been converted to a BufferGeometry:

https://gkjohnson.github.io/threejs-sandbox/mergeVertices/animated.html

And the relevant conversion code (with your suggested changes):

https://github.com/gkjohnson/threejs-sandbox/blob/master/mergeVertices/animated.html#L91

```js
loader.load( '.../examples/models/animated/sittingBox.js', function( geometry ) {

    // ...

    var bufferGeometry = new THREE.BufferGeometry();
    bufferGeometry.fromGeometry( geometry );
    bufferGeometry.removeAttribute( 'normal' );
    if ( geometry.colors.length === 0 ) bufferGeometry.removeAttribute( 'color' );

    var bmesh = new THREE.Mesh( bufferGeometry, morphMaterial.clone() );
    buffMixer = new THREE.AnimationMixer( bmesh );
    buffMixer.clipAction( geometry.animations[0] ).setDuration( 10 ).play();
    scene.add( bmesh );

});
```

And of course the mixer is updated in the render loop. You can see that the BufferGeometry version is static, though. I feel like I'm missing something obvious but I'm not sure what.

donmccurdy commented:
@gkjohnson The problem is that the animation tracks specify the morph targets they affect by name, and those names are lost when converting Geometry to BufferGeometry. Compare `geomMixer._actions[0]._propertyBindings` and `buffMixer._actions[0]._propertyBindings` in your example, and you'll see that the `.buffer` property of the latter contains `NaN`. That's hard to fix (_should_ these BufferAttributes take the names of the morph targets, since BufferGeometry names come from the `.morphTargetDictionary` property on the mesh?) so I wouldn't worry about it here. Apparently the code worked for me because I was converting to BufferGeometry and then exporting/importing the result with GLTFExporter/GLTFLoader, which flattens `clip.tracks` into a single track and inherently avoids this whole issue.

gkjohnson committed:
Support merging morphAttributes

gkjohnson commented:
Thanks @donmccurdy! I just pushed the changes that add support for merging the morph attributes and tested that the lengths of the `position` morph attribute arrays are the same as the length of the regular `position` attribute. Unfortunately I still wasn't able to save the GLTF with animations, otherwise I'd be happy to test it there. Here was my process:

```js
loader.load( '.../animated/sittingBox.js', function( geometry ) {
   
   var bufferGeometry = new THREE.BufferGeometry();
    bufferGeometry.fromGeometry( geometry );
    bufferGeometry.removeAttribute( 'normal' );
    if ( geometry.colors.length === 0 ) bufferGeometry.removeAttribute( 'color' );

    var bmesh = new THREE.Mesh( bufferGeometry, morphMaterial.clone() );
    var gltf = (new THREE.GLTFExporter()).parse( bmesh, { animations: goemetry.animations } );
    
    // save the file   

} );
```

I also realized that technically every morph attribute value should be included in the attribute hash, as well because it could be that two vertices start out near eachother but separate over the course of the animation.

I notice that the `THREE.Geometry.mergeVertices` doesn't handle morphTargets at all so it's a step up from that, but it's definitely a caveat with this function.

A slightly related aside -- I've read a few things that sound like `THREE.Geometry` is eventually going to be removed from `THREE`? Is that the case? Is there a timeline for that?

donmccurdy commented:
> Unfortunately I still wasn't able to save the GLTF with animations, otherwise I'd be happy to test it there.

Sorry yeah, the conversion to glTF depends on new features in GLTFExporter that I haven't submitted a PR for yet. But you should be able to test with a quick hack:

```js
bmesh.morphTargetDictionary = mesh.morphTargetDictionary;
for ( var key in bmesh.morphTargetDictionary ) {
  var index = bmesh.morphTargetDictionary[key];
  bmesh.geometry.morphAttributes.position[+index].name = key;
}
```
This change fixes your earlier demo (https://gkjohnson.github.io/threejs-sandbox/mergeVertices/animated.html) for me. It's not really the right way to convert the geometry, but I think it demonstrates your mergeVertices() implementation is correct.

gkjohnson commented:
Thanks -- I've updated example page to show two a before and after vertices have been merged on the animated BufferGeometry:

https://gkjohnson.github.io/threejs-sandbox/mergeVertices/animated.html

gkjohnson committed:
Fix incorrect variable reference. "attribute" -> "oldAttribute"

gkjohnson committed:
Merge branch 'dev' into optimize-buffer-geometry-indices

donmccurdy reviewed (changes requested):
Suggestions:

- [ ] Move `.mergeVertices` to BufferGeometryUtils
  - I can't think of a compelling reason it needs to be in the BufferGeometry class, so inclined to leave it out.
- [ ] Consider leaving `InterleavedBufferAttribute` clone and copy methods for another PR
  - Are they needed in this one? I don't see where, if so. We've discussed these methods before (https://github.com/mrdoob/three.js/issues/15395, https://github.com/mrdoob/three.js/pull/12800#discussion_r155131296) and it was thorny, so separating them would simplify this review.
- [ ] (Optional) Add a parameter to `mergeVertices` so it works with points and lines
  - The fact that this requires an extra parameter seems like an argument in favor of creating MeshGeometry / PointGeometry / ... classes.
- [ ] (Optional) Basic unit tests

gkjohnson committed:
Move `mergeVertices` to `BufferGeometryUtils`

gkjohnson commented:
> - [ ] Move `.mergeVertices` to BufferGeometryUtils

Done. The function now returns a new `BufferGeometry` instance, as well.

> - [ ] Consider leaving `InterleavedBufferAttribute` clone and copy methods for another PR

Are you referring to these lines?

```js
if ( oldAttribute.isInterleavedBufferAttribute ) {

    attribute = new THREE.BufferAttribute( buffer, oldAttribute.itemSize, oldAttribute.itemSize );

} else {

    attribute = geometry.getAttribute( name ).clone();
    attribute.setArray( buffer );

}
```

You mean I should get rid of the special case for the interleaved attributes? I can add a warning log, instead.

The other two I'll leave for other PRs, I think. I'll take a look at adding unit tests but the point geometry etc seems more complicated.

You can see the new function working in the two examples in my original repo:

https://gkjohnson.github.io/threejs-sandbox/mergeVertices/
https://gkjohnson.github.io/threejs-sandbox/mergeVertices/animated.html

yomboprime commented:
This could serve as a unit test: Could you modify the `webgl_physics_volume` example? There are only two minor changes:

1 - Edit the processGeometry function to this:
```javascript
function processGeometry( bufGeometry ) {
	// Obtain a indexed BufferGeometry with merged vertices
	var indexedBufferGeom = THREE.BufferGeometryUtils.mergeVertices( bufGeometry, 3 );
	// Create index arrays mapping the indexed vertices to bufGeometry vertices
	mapIndices( bufGeometry, indexedBufferGeom );
}
```
2 - Delete the function `createIndexedBufferGeometryFromGeometry`

donmccurdy commented:
> You mean I should get rid of the special case for the interleaved attributes? I can add a warning log, instead.

You've added `.copy()` and `.clone()` methods to the `InterleavedBufferAttribute` class, but then in `mergeVertices` you've specifically avoided calling `clone()` on the interleaved attributes. So it doesn't seem like there's any need to modify `InterleavedBufferAttribute.js` here?

gkjohnson committed:
Remove the interleavedBufferAttribute and copy, clone functions

gkjohnson commented:
> You've added .copy() and .clone() methods to the InterleavedBufferAttribute class, but then in mergeVertices you've specifically avoided calling clone() on the interleaved attributes. So it doesn't seem like there's any need to modify InterleavedBufferAttribute.js here?

Ha, I completely forgot I did that (it's been awhile) -- I've removed it, now.

> This could serve as a unit test: Could you modify the webgl_physics_volume example? There are only two minor changes:

That's not a bad idea. I'll look into doing that in a bit.

Thanks!

gkjohnson committed:
Remove interleavedGeometryAttribute test

gkjohnson committed:
Use mergeVertices in physics volume example

gkjohnson committed:
add comments to example

donmccurdy commented on examples/webgl_physics_volume.html:
```diff
@@ -190,60 +191,19 @@
-				var indexedBufferGeom = createIndexedBufferGeometryFromGeometry( geometry );
-
-				// Create index arrays mapping the indexed vertices to bufGeometry vertices
+                // Create index arrays mapping the indexed vertices to bufGeometry vertices
```
nit: i think this got converted from tabs to spaces

mrdoob replied:
Yep. Line 199 too.

gkjohnson replied:
Thanks I'll get these when I get home today

gkjohnson replied:
Done! Let me know if there's anything else!

gkjohnson committed:
Fix tabs

mrdoob commented:
Thanks!

WestLangley commented:
Can you update the docs please?
